### PR TITLE
Make WeakReferenceCollection thread safe

### LINKED ISSUE #59505: CompositionContainer is not thread safe - even when constructed as thread safe
Our application makes extensive use of the System.ComponentModel.Composition libraries and uses the `CompositionContainer` from many different threads.  When the container is created, it is marked (via the constructor) as thread safe.

__We've successfully used the container like this in .NET Framework 4.6.2 for many many years__, but with a recent update to .NET 5, we're seeing some crashes.

Specifically, the path along `CompositionContainer.Compose()` does not look thread safe.  We eventually get to `RecompositionManager.GetAffectedParts`, and while [enumerating the alive items list](https://source.dot.net/#System.ComponentModel.Composition/Microsoft/Internal/Collections/WeakReferenceCollection.cs,ba0bb95f9621c62b,references), an exception is thrown because a different thread has modified that list.  The other thread is in the middle of satisfying a call to `GetExportedValue`.  Examination of the crashing stack reveals that there is no locking around the modified members.

Here's a snapshot of the stack that crashes:
image

The crash has been seen against .NET 5 (runtime version 5.0.9) and .NET 6 (runtime version 6.0.0-rc.1.21451.13).

[Here](https://gist.github.com/nick-beer/308780c9502b00cd9e82a6b0bc286249) is a (relatively) simple single file c# program that will reliably reproduce the crash we're seeing.

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #59505

### CODE REVIEW

**Comment by pedrobsaila** [comment]
There are in fact 2 options to solve the issue: Either locking or creating a snapshot of list _items while enumerating inside WeakReferenceCollection class. I prefer the first because the internal list _items class is not thread-safe. Therefore adding, removing, clearing ... would also cause the same issues in multithreaded context.

**Comment by danmoseley** [comment]
Is it feasible/does it make sense to attempt to create a test that validates the safety? it's often possible to create a test that fails fairly reliably when a lock is missing.

eg: https://github.com/danmoseley/runtime/blob/40e0738d86079b833604ba62d7a0d3a578f92d99/src/libraries/System.Text.RegularExpressions/tests/Regex.Match.Tests.cs#L1797
and there are likely others.

**Comment by danmoseley** [comment]
Thanks. @buyaa-n @joperezr own this area and can help review.

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs [hunk 1]
```diff
@@ -18,7 +25,10 @@ public void Add(T item)
                 CleanupDeadReferences();
             }
 
-            _items.Add(new WeakReference(item));
+            using (_lock.LockComposition())
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs:28]: It looks better to use the `narrow lock` as it is not changing a composition
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/CompositionLock.cs#L11-L12
```suggestion
            using (_lock.LockStateForWrite())
```

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs [hunk 2]
```diff
@@ -27,7 +37,10 @@ public void Remove(T item)
 
             if (index != -1)
             {
-                _items.RemoveAt(index);
+                using (_lock.LockComposition())
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs:40]:
```suggestion
                using (_lock.LockStateForWrite())
```

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs [hunk 3]
```diff
@@ -38,7 +51,10 @@ public bool Contains(T item)
 
         public void Clear()
         {
-            _items.Clear();
+            using (_lock.LockComposition())
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs:54]:
```suggestion
            using (_lock.LockStateForWrite())
```

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs [hunk 4]
```diff
@@ -58,18 +74,24 @@ private int IndexOf(T item)
         // Should be executed under a write lock
         private void CleanupDeadReferences()
         {
-            _items.RemoveAll(w => !w.IsAlive);
+            using (_lock.LockComposition())
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs:77]:
```suggestion
            using (_lock.LockStateForWrite())
```

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs [hunk 5]
```diff
@@ -58,18 +74,24 @@ private int IndexOf(T item)
...
+                _items.RemoveAll(w => !w.IsAlive);
+            }
         }
 
         public List<T> AliveItemsToList()
         {
             List<T> aliveItems = new List<T>();
 
-            foreach (var weakItem in _items)
+            using (_lock.LockComposition())
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs:87]:
```suggestion
            using (_lock.LockStateForRead())
```

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs [hunk 6]
```diff
@@ -38,7 +51,10 @@ public bool Contains(T item)
...
         public void Clear()
         {
-            _items.Clear();
+            using (_lock.LockComposition())
+            {
+                _items.Clear();
+            }
         }
 
         // Should be executed under at least a read lock.
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/WeakReferenceCollection.cs:60]: Says the usages of `IndexOf(T item)` also need to use ReadLock

**Comment by buyaa-n** [comment]
Thanks, @pedrobsaila, I am not familiar with the code in this are therefore not that comfortable adding so many granular locks in the collection, you know it would affect the perf and most importantly could cause a deadlock. Looking into the existing locks pattern in `ImportEngine` i see the lock used for accessing `_recompositionManager` instance https://github.com/dotnet/runtime/blob/1c7100b0c21a2af147362c7b5c61408611743ded/src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/ImportEngine.cs#L637-L641 Seems the only case where `_recompositionManager` used without a lock is https://github.com/dotnet/runtime/blob/1c7100b0c21a2af147362c7b5c61408611743ded/src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/ImportEngine.cs#L495 where the exception occurs, i think we should add the write lock over there, what you think?

**Comment by pedrobsaila** [comment]
I tried to add a narrow write lock locally to the _recompositionManager.GetAffectedParts call and it seems to be solving the issue.

File: src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/CompositionContainerTests.cs [hunk 7]
```diff
@@ -2987,5 +2988,107 @@ public class PartWithRequiredImport
...
         }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
+        public async Task ThreadSafe_CompositionContainer_Calling_Compose_And_GetExport()
+        {
+            var catalog = new AssemblyCatalog(typeof(Export1).Assembly);
+
+            for (int i = 0; i < 100; i++)
+            {
+                var container = new CompositionContainer(catalog, CompositionOptions.IsThreadSafe);
```
Comment (steveharter) [inline src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/CompositionContainerTests.cs:2999]: Without `IsThreadSafe` does this test fail (at least periodically)?

Comment (pedrobsaila) [reply src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/CompositionContainerTests.cs:2999]: Yes it fails with CompositionOptions.Default

**Comment by steveharter** [comment]
See recent comment on the original issue: https://github.com/dotnet/runtime/issues/59505#issuecomment-965813857 where I'm not sure this should be fixed.

> where the exception occurs, i think we should add the write lock over there instead, what you think? CC @steveharter @stephentoub

>> I tried to add a narrow write lock locally to the _recompositionManager.GetAffectedParts call and it seems to be solving the issue.

It seems like we should be locking the calling method instead of putting all of the locking in `WeakReferenceCollection`. Also based on this:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/CompositionLock.cs#L10-L21
in particular ```No 3rd party code will EVER get called inside the narrow lock``` if we use a narrow lock we should be sure no events are fired within that lock. However, I think it would be consistent just to use a wide lock there.

In summary, going forward: first verify we want to fix this. If so then:
- Undo the changes to `WeakReferenceCollection`
- Add a wide composition lock to `GetAffectedParts()` via ```using (_lock.LockComposition())``` OR if more consistent with other areas use a narrow lock if no events are fired.
- Add more tests to see if there are other issues with recomposition + multiple threads.

**Comment by dsplaisted** [comment]
Based on my [11 year old comment](https://stackoverflow.com/a/3252834) on Stack Overflow that @steveharter uncovered and my distant recollections, I would be hesitant to try to fix this.  I think that MEF was not designed to support being thread-safe in this manner, probably because we didn't think there was a way to do so safely and correctly.  I suspect that if we were to try to fix crashes like this that you could still end up with inconsistent compositions or container state when doing concurrent operations like this.  But it's been a long time since I've dealt with any of this so I'm not entirely sure.

**Comment by danmoseley** [comment]
Perhaps @AArnott has a historical perspective on thread safety here (?) or is this the other MEF.

**Comment by joperezr** [comment]
Also tagging @weshaggard who worked on MEF in case he knows.

File: src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/CompositionContainerTests.cs [hunk 8]
```diff
@@ -2987,5 +2988,107 @@ public class PartWithRequiredImport
...
+
+                var batch2 = new CompositionBatch(partsToRemove: new[]
+                {
+                    AttributedModelServices.CreatePart(new Export12())
+                }, partsToAdd: Enumerable.Empty<ComposablePart>());
+                container.Compose(batch);
+
+                Exception exception = await Record.ExceptionAsync(() => task);
+                Assert.Null(exception);
+            }
```
Comment (steveharter) [inline src/libraries/System.ComponentModel.Composition/tests/System/ComponentModel/Composition/CompositionContainerTests.cs:3030]: The actual compositions should be verified to see if recomposition worked as expected.

**Comment by AArnott** [comment]
My distant memory suggests I say that System.ComponentModel.Composition was never thread-safe for recomposition, and that the thread-safe flag was only to make the container thread-safe for obtaining exports across many threads. From my memory, I _think_ thread-safety in obtaining exports was good in this library.

System.Composition BTW appeared to have better thread-safety but in my rather thorough thread-safety testing of VS-MEF, I found bugs in System.Composition around thread-safety as well which AFAIK may not have ever been fixed, even in obtaining of exports (as it has no recomposition feature).

**Comment by weshaggard** [comment]
My recollection is similar to @dsplaisted and @AArnott that MEF is inherently not thread-safe, and especially not in recomposition scenarios. I would also be hesitant in taking this fix with the expectation that things will be thread safe in the library, it might help some targeted scenarios but overall I expect there to still be issues using this across threads.

**Comment by buyaa-n** [comment]
Thank you all for your valuable input, your suggestions confirm my doubt about this fix. As we all greed I am closing this PR. 

Thank you for your time @pedrobsaila
